### PR TITLE
Add extract code actions

### PR DESCRIPTION
**Pull request by jfeser**
Adds two new code actions: `Extract local` and `Extract function`. 

Extract local transforms:
```
let f x =  $x+1$ + 2
```
to 
```
let f x = 
  let new_var = x + 1 in 
  new_var + 2
```
The new variable gets placed at the tightest enclosing binder of the extracted expression.

 Extract function would transform the same code to:
```
let new_fun x = x + 1
let f x = new_fun x + 2
```
The new function will be inserted directly before the enclosing structure item. Any free variables in the expression will be passed as arguments.

Just wanted to get opinions on whether this is a useful feature, and whether these behaviors make sense.

TODO:
- Should extract local move the expression to the loosest binder rather than the tightest?
- Extract function doesn't handle free variables with nontrivial `Longident`s.
- Add tests.

### CODE REVIEW

**Comment by rgrinberg** [comment]
The feature looks useful to me.

> Should extract local move the expression to the loosest binder rather than the tightest?

I would expect the tightest. But perhaps you could have two extract actions for the different behaviors.

Or perhaps we could have different refactorings for hoisting bindings up.

> The new function will be inserted directly before the enclosing structure item.

I would expect it to stay close to the scope where it's being used personally.

**Comment by rgrinberg** [comment]
Btw, can you try splitting this code action into a discovery/resolve phase and evaluate if it helps performance a little. I'm getting a little worried about all these code actions clogging up the server.

**Comment by jfeser** [comment]
I have a branch (https://github.com/jfeser/ocaml-lsp/tree/tracing) with some extra tracing. From the traces I've looked at, code actions aren't a performance concern. How have you been evaluating performance/where have you seen performance problems?

File: ocaml-lsp-server/test/e2e-new/code_actions.ml [hunk 1]
```diff
@@ -599,3 +599,98 @@ let f (x : t) = x
...
+      cnum := 0)
+    else incr cnum
+  done;
+  Position.create ~character:!cnum ~line:!lnum
+
+let parse_selection src =
+  let start_pos =
+    match String.index src '$' with
+    | Some x -> x
+    | None -> failwith "expected a selection opening mark"
```
Comment (rgrinberg) [inline ocaml-lsp-server/test/e2e-new/code_actions.ml:619]: Is it possible the user can encounter this error?

1. If yes, then it should be a Jsonrpc error.
2. if not, it should it be a code error.

Comment (jfeser) [reply ocaml-lsp-server/test/e2e-new/code_actions.ml:619]: This is part of the test harness. It gets raised when the test input doesn't have $ markers for the selection. Do you want me to use a different exception?

Comment (rgrinberg) [reply ocaml-lsp-server/test/e2e-new/code_actions.ml:619]: Oh never mind. It's fine.

**Review by rgrinberg — APPROVED** [review]

**Comment by 3Rafal** [comment]
@rgrinberg , unit tests for this PR are broken. CI fails on 4/7 checks

**Comment by rgrinberg** [comment]
@jfeser could you take a look?

**Pull request merged**